Implement presence monitoring

pbui opened the pull request:
This feature implements a crude form of user presence by using the CHAN
MODE IRC command to denote if a user is active (+v / voiced) or away (-v
/ devoiced).  The use of voice and devoice to denote presence is based
on how bitlbee communicates status in its IRC bridge.

- To receive presence updates, we must subscribe to each user we are
interested in; unfortunately there is no way to know which channel the
IRC user is viewing, so we simply subscribe to all users.  This
exhaustive approach is expensive, but it is infrequent.

- Likewise, to communicate a change in user presence, we must send a
CHAN MODE command to each channel the user is in.  Again this is
expensive, but it is infrequent and the only way since IRC does not
directly support the concept of user presence.

- No additional bookkeeping is required since the ircUser class already
keeps track of all users in its channelNicks member.

- Currently, although batch_presence_aware is set, it appears to do
nothing (presence updates still occur one user at a time).

- On start up, there is a huge flood of traffic since the Slack server
will send updates for many (if not all) of the users (and the bridge
will then send CHAN MODE commands).

- This feature is disabled by default and is only enabled by the
"presence" password argument (sent by the IRC client after the Slack
token).

adsr commented on lib/irslackd.js:
```diff
@@ -726,6 +745,22 @@ class Irslackd {
+      });
+    });
+    console.log('Subscribing to presence of', userIds.size, 'users');
+    ircUser.slackRtm.subscribePresence(Array.from(userIds));
```
I think we should also call `subscribePresence` when someone new joins a channel the user is in. Either that or possibly there is a 'new user' event that we can subscribe to. Not sure if that event exists.

Looks great otherwise. Will merge with unit tests. Thanks again!

pbui replied:
Would the [team_join](https://api.slack.com/events/team_join) make sense for this?

Otherwise, I can call `updatePresenceSubscriptions` at the end of `onSlackMemberJoinedChannel` (this should be OK since most people don't constantly join and leave channels on Slack).

pbui replied:
I guess one problem with `team_join` is that the new member may not be in a channel the user is in and thus will not get a subscription.  This is OK, but if the new member later joins a channel the user is in, it will still not get a subscription.  Therefore, it is probably best to do the call at the end of `onSlackMemberJoinedChannel`. 

Let me know what you think.

adsr replied:
> I guess one problem with team_join is that the new member may not be in a channel the user is in and thus will not get a subscription.

Is that documented somewhere? Reading over the docs it seems like you are free to subscribe to any user in the workspace. I ask because doing this in `team_join` would be cleaner and more efficient.

pbui replied:
This is an implementation issue on our side.  The `updatePresenceSubscriptions` function only adds users that are in `channelNicks` (ie. nicks that were already in the room when the IRC client joins or nicks that have joined the room since the IRC client has joined).

It is possible that on `team_join` that the new member is not in any channel that the IRC client is in, and thus would never be added to `channelNicks`.

Practically speaking, this is probably not true since I believe everyone has to be in `#general` or the equivalent and thus a `onSlackMemberJoinedChannel` would be triggered.  Thus adding `updatePresenceSubscriptions` to the end of `onSlackMemberJoinedChannel` would take care of this.

I have been running the server with modification to `onSlackMemberJoinedChannel` for a few days and it seems fine (a few disconnects here and there, but presence monitoring works).

That said, I agree that using `team_join` would be more efficient.  I'm still open to either approach.

adsr replied:
Ah, I see what you mean. Instead of using `channelNicks` can we use the keys of `slackToIrc` and pick out anything that begins with a 'U'? If so I'd put this behind a method `IrcUser::getAllSlackUserIds` or something so that we can easily change the implementation down the line if desired. (In the future it may be desirable to split `slackToIrc` and `ircToSlack` maps into separate maps per object type (channel, user, bot, team, etc)). Then the `team_join` handler can simply invoke `mapIrcToSlack` and that should keep things updated. What do you think?

pbui commented:
@adsr I've updated the code:

1. We now have a `IrcUser::getAllSlackUserIds()` method which uses the `slacktoIrc` map to construct a list of all Slack UserIds.

2. The new `updatePresenceSubscriptions` method now uses the `IrcUser::getAllSlackUserIds()` method.

3. We now have an `onSlackTeamJoin` event handler that adds the new team member via `IrcUser::mapIrcToSlack()`; we call `updatePresenceSubscriptions` after we add this new team member.

4. The last major modification is that I moved initial updating presence subscribitions to after all users, teams, and channels have been refreshed.  Additionally, I use a timeout to only call the `updatePresenceSubscriptions` method after 5 seconds.

The reason for this ordering and delay is that the IRC client is notified of user presence via CHAN MODE messages.  If the client is not yet in those channels, then any updates are discarded.  By performing the presence subscription upate after we have refreshed the channels (and some timeout to give the client time to process), we make it less likely that we send CHAN MODE messages before the IRC client is ready.

I chose a 5 second timeout because that appear to be enough time for my use case, but we can always make it higher to be safe (all it does is delay the processing of presence events).

If there is a better way to handle this race condition, please let me know.

adsr commented:
Sounds good. I think one way to handle the race condition you mention in bullet 4 is to keep a map of presence keyed by user (`IrcUser.ircNickPresenceMap` or something). We can then update this map in `onSlackPresenceChange` and only issue a `MODE` command if `ircUser.isInChannel(...)`. Later in `sendIrcChannelJoin`, when we're sending `353`s, we can check the map and send `+nick` instead of `nick` if the user is active. Take a shot if you feel like it, or if not, file an issue for later and I'll merge it as-is after you add a test.

Thanks again @pbui.

pbui commented:
I've followed your suggestion and have created a set called `IrcUser.ircNickActive` which contains all the currently active users.  This is updated in `onSlackPresenceChange` by removing users that are away and adding users that are active.  Likewise, we check the `ircUser.ircNickActive` set in the `sendIrcChannelJoin` method to prepend a `+` to the `nick` if the `nick` is considered active (as you described).

Additionally, when issuing a `CHAN MODE` we now check if both the ircUser and the changed member are in the channel we are checking before sending the IRC command.

In my current live-testing it appears to work great and is stable.  I will let it run overnight and then work on some unit tests tomorrow.

pbui commented:
I've added a unit test for testing `onSlackPresenceChange`.

pbui commented:
I also have a diff for the wiki that documents the two changes:

https://gist.github.com/pbui/186ba9013474b3ab38b7c68abdb988be
